app/src/behaviors/behavior_smart_interrupt.c

```diff
@@ -0,0 +1,288 @@
+}
+
+static void clear_smart_interrupt(struct active_smart_interrupt *smart_interrupt) {
+    smart_interrupt->is_active = false;
+}
+
+struct active_smart_interrupt active_smart_interrupts[ZMK_BHV_MAX_ACTIVE_SMART_INTERRUPTS] = {};
+
+static struct active_smart_interrupt *find_smart_interrupt(uint32_t position) {
+    for (int i = 0; i < ZMK_BHV_MAX_ACTIVE_SMART_INTERRUPTS; i++) {
+        if (active_smart_interrupts[i].position == position &&
+            active_smart_interrupts[i].is_active) {
+            return &active_smart_interrupts[i];
+        }
+    }
+    return NULL;
+}
+
+static int new_smart_interrupt(uint32_t position,
+                               const struct behavior_smart_interrupt_config *config,
+                               struct active_smart_interrupt **smart_interrupt) {
+    for (int i = 0; i < ZMK_BHV_MAX_ACTIVE_SMART_INTERRUPTS; i++) {
+        struct active_smart_interrupt *const ref_smart_interrupt = &active_smart_interrupts[i];
+        if (!ref_smart_interrupt->is_active) {
+            ref_smart_interrupt->position = position;
+            ref_smart_interrupt->config = config;
+            ref_smart_interrupt->is_active = true;
+            ref_smart_interrupt->is_pressed = false;
+            ref_smart_interrupt->first_press = true;
+            *smart_interrupt = ref_smart_interrupt;
+            return 0;
+        }
+    }
+    return -ENOMEM;
+}
+
+static bool is_other_key_shared(struct active_smart_interrupt *smart_interrupt, int32_t position) {
+    for (int i = 0; i < smart_interrupt->config->shared_key_positions_len; i++) {
+        if (smart_interrupt->config->shared_key_positions[i] == position) {
+            return true;
+        }
+    }
+    return false;
+}
+
+static bool is_layer_shared(struct active_smart_interrupt *smart_interrupt, int32_t layer) {
+    for (int i = 0; i < smart_interrupt->config->shared_layers_len; i++) {
+        if (smart_interrupt->config->shared_layers[i] == layer) {
+            return true;
+        }
+    }
+    return false;
+}
+
+static int on_smart_interrupt_binding_pressed(struct zmk_behavior_binding *binding,
+                                              struct zmk_behavior_binding_event event) {
+    const struct device *dev = device_get_binding(binding->behavior_dev);
+    const struct behavior_smart_interrupt_config *cfg = dev->config;
+    struct active_smart_interrupt *smart_interrupt;
+    smart_interrupt = find_smart_interrupt(event.position);
+    if (smart_interrupt == NULL) {
+        if (new_smart_interrupt(event.position, cfg, &smart_interrupt) == -ENOMEM) {
+            LOG_ERR("Unable to create new smart_interrupt. Insufficient space in "
+                    "active_smart_interrupts[].");
+            return ZMK_BEHAVIOR_OPAQUE;
+        }
+        LOG_DBG("%d created new smart_interrupt", event.position);
+    }
+    LOG_DBG("%d smart_interrupt pressed", event.position);
+    stop_timer(smart_interrupt);
```

Comment:
Why are we stopping this timer in the press? Smells like missed clean up elsewhere if this is needed.